### DISCUSSION TITLE
Theme for charts

### DISCUSSION THREAD

**Discussion by Eralmidia** [IgniteUI/igniteui-angular#10449] [Q&A]
## Question  
I know there is a repository for the charts, but I believe the theming support were created by you guys, and since you are generally quick to respond, I'm hoping you can give me some pointers on themes for charts.

To put it simply, I don't understand how I'm supposted to style my charts at all. Since we upgraded from igx 10 to 12, I can't seem to get my charts to apply the correct brushes. Formerly, we generally fed the brushes to the brushes inputs of either the charts or series, but this no longer seems to work, because something in our theme.scss is overriding it.

My next attemt was setting up a theme, but this doesnt seem to work either, because the color palette supplied to the igx-theme seems to be overriding any colors set. So, for instance, I have the following set up for my applications dark theme:

![image](https://user-images.githubusercontent.com/12936286/132861332-a62d9d4d-c2a7-4af7-9242-fc0a8a71df0d.png)

As far as I understand, this should now give the first three series blue, pink and red colors. Looking at a data chart however, I get this:

![image](https://user-images.githubusercontent.com/12936286/132861509-c53a9382-5315-41cf-9c1b-622830a97a3f.png)

The yellow color is actually provided by the palette. If I remove `@include igx-theme($color-palette);` from the theme
![image](https://user-images.githubusercontent.com/12936286/132862153-c8609a50-59ef-4155-a900-0c66eda8237a.png)

the colors of the theme will suddenly start working:
![image](https://user-images.githubusercontent.com/12936286/132862353-d03f0066-8f66-42e4-bd4a-a93b56b0a1e3.png)

This will however mess with how other components look, like for instance the time picker, so I'm assuming this is not how I'm suppose to solve this:
![image](https://user-images.githubusercontent.com/12936286/132862508-a8e20ed9-2ac3-4b43-b9a1-dd793788dbb4.png)

It does seem like the theme is actually applied, because if I add forinstance a color the the title, like so:
![image](https://user-images.githubusercontent.com/12936286/132863560-85da7ca4-a159-48b0-8da1-c5faedeeaa72.png)
this is actually displayed as expected:
![image](https://user-images.githubusercontent.com/12936286/132863639-46bb1fe4-e68b-4f9e-b7a7-cfb65d7f49c6.png)

Also, in the docs, I can see text color for title and sub title. How exactly do I set the label color?

I'm attaching the full theme.scss as well.
[theme.zip](https://github.com/IgniteUI/igniteui-angular/files/7144136/theme.zip)

 * igniteui-angular version: 12

**Comment by Eralmidia** [comment]
Anyone?

**Comment by simeonoff** [comment]
Hi @Eralmidia we will look into this issue next week.

**Comment by Eralmidia** [comment]
@simeonoff Which next week? 😛

**Comment by simeonoff** [comment]
> @simeonoff Which next week? 😛

Hi, just added it to this sprint.

**Comment by simeonoff** [comment]
@gmurray81 can you chime in? I'm testing the charts in Ignite UI for Angular v12.2.0 and some styling properties no longer work. I tried setting the range brushes for the linear gauge via both property binding and custom CSS properties, alas none of these methods work anymore. Obviously this affects the theming system.

**Comment by gmurray81** [comment]
I'll take a loot at this on Monday. I fixed some stuff recently in this area, but that actually had to do with the other way around where we had a regression specific to ES2015+ where the css properties would always take precedence. But that doesn't seem like what is being described here, so will look into this.

**Comment by gmurray81** [comment]
@Eralmidia 
you can use this css custom prop for the axis labels:
`--axis-label-text-color: orange;`

https://stackblitz.com/edit/github-dmjefr?file=src%2Fapp%2Fscatter-point-chart%2Fdata-chart-scatter-point-chart.component.scss

That all still works if I upgrade the packages to latest, so I'll have to see what you were having issues with specifically @simeonoff

**Comment by gmurray81** [comment]
Or are you talking about an internal package version?

**Comment by gmurray81** [comment]
If you are talking about the latest internal versions, maybe that explains what you are seeing, depending on the sample. I fixed an issue where in some cases the css would always take precedence, even over a local value set by the user. That's been resolved. So maybe the sample you were using has some local values?

Another possibility though, is perhaps some specific  scenarios are getting confused and thinking that an internal default constitutes a user provided local value. In which case that will just need to be resolved.

**Comment by Eralmidia** [comment]
Any news on this? This is kinda blocking our release now, and becoming a major issue for us.

**Comment by gmurray81** [comment]
@Eralmidia did you try my suggestion above?

```css
--axis-label-text-color: orange
```

**Comment by Eralmidia** [comment]
@gmurray81 @simeonoff 

No unfortunately it's not working. It seems to be the same issue as the color of the columns, it doesn't listen to the theme. The title however works. So the following:
![image](https://user-images.githubusercontent.com/12936286/139434023-703fd0dc-2abd-4299-9d88-f4a82a18b5f4.png)

results in the following:
![image](https://user-images.githubusercontent.com/12936286/139434094-c9fb8c94-2438-4dd9-addb-e348f0b2e0a0.png)

Both label and brushes seems to be overwritten by the palette:
![image](https://user-images.githubusercontent.com/12936286/139434438-3efa58d2-fce2-48fe-86c1-db0bcb4452e7.png)

![image](https://user-images.githubusercontent.com/12936286/139436024-1f3307d2-dbe8-4e39-86c6-d32ddea0733c.png)


If I remove the palette the colors from the theme (or brushes inputs on the charts for that matter) starts working:

![image](https://user-images.githubusercontent.com/12936286/139435061-dcd03a9e-431e-4ce0-958e-e5b0e9c3100b.png)

![image](https://user-images.githubusercontent.com/12936286/139434886-999a710e-08f9-4f24-82fc-bdcf3eea1645.png)

But as mentioned, lots of other things break in the CSS in other components if I remove that line, so I'm assuming that's meant to be there.

@kdinev Any chance we can bump this up from low priority, this is kinda GUI-breaking for our application?

**Comment by gmurray81** [comment]
hmm... those properties were working just fine outside of the main scss theme, so maybe something in the theme is taking higher precedence (higher specificity)? you could try using important to see if it behaves. I'm not really familiar with the structure of the theme, so I'll let @simeonoff  investigate that aspect of it.

Let me try loading the full theme in the sample I sent you and see if the behavior is the same. What's the exact package version you are using @Eralmidia? So that I can make sure I'm comparing like for like.

**Comment by gmurray81** [comment]
I updated this [stackblitz](https://stackblitz.com/edit/github-dmjefr?file=src%2Fstyles.scss) to import the theme and, presuming I did that correctly, it seems like I can still set the axis title color using the css property I mentioned. But the theme does also try to set this, so maybe its a specificity issue?

**Comment by Eralmidia** [comment]
Setting the properties as important didn't do anything. But yes, a specificity issue definitely seems like the most likely cause

**Comment by Eralmidia** [comment]
Hmm, actually, maybe its not?
![image](https://user-images.githubusercontent.com/12936286/139438941-97c2eb06-f7d5-4b58-b085-7479ceb81116.png)
 Doesn't seem like anything is overriding the label color?

**Comment by Eralmidia** [comment]
And there doesn't seem to be anything else trying to set this property:
![image](https://user-images.githubusercontent.com/12936286/139439464-5ee8c143-75f2-4eaf-b044-52a26ed0ef0e.png)

**Comment by gmurray81** [comment]
So the title is working in that scenario?
Seems red. But the axis color isnt?
There's another possibility at play here. Where we give multiple ways to address the color of sub-elements of the chart. 
For example, you can often directly address them --label-text-color or address them by sub element --axis-label-text-color or be very specific --category-x-axis-label-text-color.

If you are using a different mode to address them than the theme, its possible both properties are live and it may end up being last one in wins. Let me see if I can see what is being set for the axis labels.

**Comment by Eralmidia** [comment]
Yes, title is working fine, the label is not. But honestly, the label color is a secondary issue. The main issue really is that the brushes doesn't work either. They use the color of the palette and nok brushes set in the theme or from the input on the chart.

**Comment by gmurray81** [comment]
oddly, though, --axis-label-text-color is working just fine in my sample. Perhaps we aren't testing the same version?

**Comment by Eralmidia** [comment]
I'm testing "@infragistics/igniteui-angular": "12.1.7", and "igniteui-angular-charts": "12.1.1",

**Comment by Eralmidia** [comment]
But I've had this issue ever since we upgrade the igniteui-angular package from major 10 to 12

**Comment by gmurray81** [comment]
ok, I was using 12.1.6, let me update, maybe there were changes to the theme. Or its possible stackblitz is getting in the way here.

**Comment by Eralmidia** [comment]
Yep, I can downgrade to 12.1.6 also if you prefer, but I guess latest is the best.

**Comment by Eralmidia** [comment]
Also, I attached the full theme file in the first post if that is any help

**Comment by gmurray81** [comment]
Maybe it has more to to with something else in that theme file than anything to do with the version, if that's not the standard theme.

**Comment by Eralmidia** [comment]
No, I'm not sure. Maybe @simeonoff can spot something which is off in the theme file. Honestly we set this file up years ago, so maybe somehting has changed which we need to tweak, but I have no idea what if that's the case.

**Comment by simeonoff** [comment]
> Maybe it has more to to with something else in that theme file than anything to do with the version, if that's not the standard theme.

As I explained above the theme properties for the brushes don't work because the custom CSS properties for them are no longer being read by the component. I even tried binding directly to the HTML prop and the brushes still didn't change their appearance.

**Comment by gmurray81** [comment]
they are being read @simeonoff which I can demonstrate in isolation.
I think the issue is more that there are conflicts in the theme because some of the custom css props aren't being specific enough and are colliding. For example:
![image](https://user-images.githubusercontent.com/4264873/139446612-1835bd3a-3f37-4fb5-80db-758fcd0afa2e.png)

here brush and brushes are being set for the chart.
brush is winding up setting brush for each series contained in the chart. setting a brush on each series will take higher precedence than setting the brushes on the chart itself, making it seem like brushes isn't in effect.

the setting for brush, here is actually directed at sparkline I think, rather than the chart, so it should really have been:

--sparkline-brush 

to avoid the conflict.

**Comment by gmurray81** [comment]
This also wouldn't happen if all of these css props werent just being applied to the entire root by the theme, rather than at the classes we provide on each element.

**Comment by gmurray81** [comment]
or maybe you are using classes but there's a leak of this particular property? And maybe some others? Haven't dug into it.

**Comment by gmurray81** [comment]
this style of rule that I get out of the above theme file is odd, @simeonoff ....

.dark-theme, .dark-theme igx-linear-gauge {

so its applying to anything with dark-theme applied... is it the intent that all the rules should be applied to all the components or not? if they are all supposed to be in effect, I think the properties that are being targetted may need to be chosen with more care to avoid collisions. But if the rules are only supposed to apply to the appropriate components, then collisions on css custom property names are ok

**Comment by gmurray81** [comment]
@simeonoff perhaps it has something to do with how the theme in the original post here is being constructed?

When I compare to that stackblitz where I'm loading the IG theme, only chart related rules are being applied, and the issue isn't apparent.

**Comment by gmurray81** [comment]
I updated that [stackblitz](https://stackblitz.com/edit/github-dmjefr?file=src%2Fstyles.scss) @simeonoff  to show how the brushes/markerbrushes/outlines work just fine. I think the issue being hit here is more that all the rules are being applied to all the components based on how that theme is generated, and there are some collisions on property name. These can be qualified to avoid collision by putting the component name in hte css prop name e.g. --sparkline-brush rather than --brush.

But if the issue is just that the rules shouldn't be applied to all components then maybe thats the problem to address.

**Comment by gmurray81** [comment]
@simeonoff if I go in and change the sparkline scss so that it sets 
sparkline-brush: $brush,

instead of brush, it evades the issue, but there are likely a bunch of other collisions.
did something change where its suddenly  applying all the rules to all the components?

**Comment by gmurray81** [comment]
note, I'm only seeing that behavior where all the rules are applying if I apply the dark-theme class to a container of the chart

```html
<div class="dark-theme">
...
</div>
```

maybe that's not the intended use case?
I don't see that behavior if I don't specific any theme related class.

**Comment by gmurray81** [comment]
the long and the sort of it is, that if there are valid scenarios where all the rules are going to be applied to the chart, @simeonoff  is that I think you are better off using qualified names that start with the type name.

**Comment by gmurray81** [comment]
actually, its seeming like with the built in themes I don't get all the rules applied to every component. I think maybe that's just an artifact as to how @Eralmidia 's theme file is structured, essentially emitting all the rules associated with dark-theme, etc.

Is this supposed to be valid @simeonoff ? If so, the css custom props should be adjusted to use versions that don't collide between the components, if not, then could you suggest an alternate way to construct that theme?

**Comment by gmurray81** [comment]
@Eralmidia 
to avoid this specific issue, you can exclude the sparkline's styling from the theme you are generating:
```scss
.dark-theme {
    background: $dark-theme-bg;
    @include igx-theme($palette: $color-palette, $exclude: ('sparkline'));
    @include igx-banner($banner-dark-theme);
```

But because all the rules are getting mushed onto every component by these themes, there might be other conflicts.

**Comment by gmurray81** [comment]
@simeonoff 
In order for building these themes to work in the way that @Eralmidia wants, where an extra class is being applied, I think this mixin:
```scss
/// Add theme colors to a scope.
/// @access public
/// @param {map} $theme - The component theme to be used
/// @requires {mixin} css-vars-from-theme
@mixin igx-css-vars($theme, $scope: null, $prefix: null) {
    $s: map-get($theme, 'selector');
    $n: map-get($theme, 'name');
    $name: if($scope, $scope, $s or $n);

    @if not(legacy()) {
        @if is-root() {
            #{$name} {
                @include css-vars-from-theme($theme, $prefix);
            }
        } @else {
            & #{$name} {
                @include css-vars-from-theme($theme, $prefix);
            }
        }
    }
}
```

would have to be doing

```scss
& #{$name} {
```

instead of the current:

```scss
&, 
#{$name} {
```

IOW not an or operation. But I'm guessing there is some way you have this mixin defined this way for elsewhere?
It just seems a bit odd that I think @Eralmidia  is intending to produce rules like this:

```css
.dark-theme .igx-datachart {
...
}
```

and is instead getting rules like this:

```css
.dark-theme, .dark-theme .igx-datachart {
...
}
...

so that anything with .dark-theme will get _everything_

**Comment by Eralmidia** [comment]
@gmurray81 Thanks for all the investigation. Testing excluding sparkline partially worked, in that the brushes are now respected. But the label and strip colors are still off:

![image](https://user-images.githubusercontent.com/12936286/139552061-c251b18e-fa67-4c6d-be8a-ba80d2a26d76.png)

**Comment by gmurray81** [comment]
It is possible that something else is colliding with those then. Let me check your example. Hopefully @simeonoff  will be able to chime in as to what the expectation is in building the theme the way you are trying to.

My expectation would be given my limited knowledge of scss is that given how you are declaring those themes that you are expecting it to add dark-theme as an ANDed class with the classes in the theme's rules, rather than as ORed with the classes in the theme's rules which is the whole issue, since it is resulting in every rule applying to every component.

**Comment by Eralmidia** [comment]
Actually it was @simeonoff who told me to build it this way, 3 years back:
https://github.com/IgniteUI/igniteui-angular/issues/1384#issuecomment-391680477

And it has been working fine up until recently.

**Comment by gmurray81** [comment]
It's possible that some of the composition functions that build the theme. changed.. Or that maybe what changed is that more of the components are themed and so there are some colliding property names. You could try comparing the css rules that were emitted in the version that was working and today and that might shed some light on it.

The key to the other properties working is to make sure to set the same properties that the theme is setting. 
Setting --axis-title-text-color isn't working because the theme sets the more general --title-text-color and the two are fighting. Maybe we can try to improve the css extraction we do for the canvas based components so that we will favor the more specific variation (including the component type in the css property name) over the general property name if both are specified. In the meantime, its best to look at the computed properties on the element and make sure to override the matching property names:

![image](https://user-images.githubusercontent.com/4264873/139555250-aef7f943-53b0-477e-85ae-fc40053b422a.png)

```scss
.ig-data-chart {
  --plot-area-background: red !important;
  --title-text-color: red !important;
  --label-text-color: red !important;
  --category-axis-base-strip: green !important;
}
```

These properties are working fine for me. I used !important because the ones coming from the theme have higher specificity. Note though, that I had to use --title-text-color and --label-text-color because these are what the theme was using. It should probably be using more specific variations of these properties to make sure its targeting the axes. For the strip, notice I used category-axis-base-strip because if I just used strip, it would target the y axis too. But if you use other types of horizontal axis in the chart you would have to target them also. I hope this helps!

**Comment by gmurray81** [comment]
Note, I was initially having a hard time testing this, because I think that the way that angular is doing hot module reloads for the scss could be playing games with us. For the canvas based components we need to explicitly extract the CSS custom props we are looking for and need to be notified if the css environment around us changes. I suspect the tooling is injecting new CSS into the page in such a way that we don't see it, sometimes, even if you are trying to hard reload. Seems more reliable to open a new tab. If there's some way for us to listen for the hot reload system changing the CSS environment, we could make sure we respond, but I'm not sure if we'll have the right hooks for that.

**Comment by Eralmidia** [comment]
Looking at the CSS, for data chart there only seems to be these:
![image](https://user-images.githubusercontent.com/12936286/139556230-d65a2539-3e94-4cd1-b2b2-7bd3b1f8b677.png)
 I'm sure I'm just my lack of knowledge of the inner workings of the chart, but there doesn't seem to be any mention of a label property, so I'm not really sure how to identify what to override in order to fix the label and the strip colors

**Comment by gmurray81** [comment]
it may have been that they were actually being set for a different component, and its a collison. The easiest way to parse it is to switch to the computed property view in the f12 tools rather than the style view. And see what properties are in effect. It may be that those properties were coming from pie chart or category chart style.

Again, part of the issue here is that because the rule selector is:

```css
.dark-theme, .dark-theme igx-data-chart {
```

and all of the rule selectors look that way. It means that all of the rules are getting set for anything that has .dark-theme, not just the components that those rules are actually meant to apply to. So you have rules that are trying to set properties for pie-chart that are getting applied to data-chart and vice versa. This doesn't seem like its the desired outcome, but if it is, collisions could be avoided by adjusting the property names to scope them to the components.

**Comment by gmurray81** [comment]
![image](https://user-images.githubusercontent.com/4264873/139557401-8342d93e-b6b0-4bfe-8d96-fa1567d1102e.png)

`--title-text-color` is coming from quite a view places, for example.

This doesn't happen when you run `igx-theme` at the root of your scss file. When you run at the root of your file each rule is properly scoped to the component it's supposed to target. Its just when you do something like:

```scss
.dark-theme {
  @include igx-theme($palette: $color-palette, $exclude: ('sparkline'));
...
}
```

That it ends up generating selectors for `.dark-theme OR (.dark-theme AND [component class])` for the rules, which creates the issue.

**Comment by gmurray81** [comment]
I have to assume this is probably incorrect behavior, to use the OR, since your general expectation when nesting in SCSS would be AND behavior, as far as I'm aware, but I'm not super familiar with scss. But maybe there's a good reason why this kind of combination was chosen over the more expected behavior.

**Comment by Eralmidia** [comment]
Hmm, seems like it's the slider then
![image](https://user-images.githubusercontent.com/12936286/139557823-dacf53a7-6ef2-40da-bb0c-a0d8a60f3107.png)

**Comment by Eralmidia** [comment]
Ok, so setting a color for the slider label color worked, so its defintately a collision. So seems like this will be a good band aid for the problem.

**Comment by gmurray81** [comment]
@Eralmidia awesome, glad to hear.

**Comment by Eralmidia** [comment]
@gmurray81 Minor thing: I can't seem to find a variable for the strip-color in the theme docs (https://www.infragistics.com/products/ignite-ui-angular/docs/sass/latest/#function-igx-data-chart-theme). How do you set that? And also, just in case there are some issues with the other components, which css var would I need to override for this?

**Comment by gmurray81** [comment]
Sorry, to clarify my language above, by AND I mean the rule would apply to 
igx-data-chart which is ALSO within a container with dark-theme class
rather than by OR I mean the rule would apply to igx-data-chart OR a container with dark-theme class.

Its the former behavior we want, not the latter.

@Eralmidia those docs may only contain the properties that the scss mixin for creating data chart themes supports. But the chart supports many more custom css properties than that exposes. Basically you can take any visual related property on the chart and convert it to spinal case to set it as a custom css property.

In the case of the strip color, there is a property on Axis called strip, so you can set it as:

```css
--strip: red;
```

but you can also specify the containing type name to resolve ambiguities:

```css
--axis-strip: red;
```

furthermore there are a lot of different types of axes in the chart. So you potentially want to set the strip color for a particular axis type:

```css
--category-x-axis-strip: red;
```

so you can resolve ambiguities by pre-pending the target type to the property name.

**Comment by gmurray81** [comment]
to see what visual properties are available on the chart to configure its look and feel, you can check the API doc for the chart.

**Comment by Eralmidia** [comment]
Thanks @gmurray81 I assumed things were available in the theme docs like in the other components, but I'll try to locate the chart vars then 👍
